go/merkle/types.go

```diff
@@ -0,0 +1,175 @@
+package merkle
+
+import (
+	"crypto/sha512"
+
+	"github.com/keybase/client/go/msgpack"
+	"github.com/keybase/client/go/protocol/keybase1"
+	"github.com/keybase/client/go/sig3"
+	merkletree "github.com/keybase/go-merkle-tree"
+	"github.com/pkg/errors"
+)
+
+type TreeSeqno int64
+
+type EncodingType byte
+
+const (
+	EncodingTypeBlindedSHA256v1     EncodingType = 1 // p = HMAC-SHA256; (k, v) -> (k, p(p(k, s), v)) where s is a secret unique per Merkle seqno
+	EncodingTypeBlindedSHA512_256v1 EncodingType = 2 // p = HMAC-SHA512-256; (k, v) -> (k, p(p(k, s), v)) where s is a secret unique per Merkle seqno
+)
+
+const CurrentEncodingType = EncodingTypeBlindedSHA512_256v1
+
+const MaxChildrenPerLeaf = 2
+
+func GetTreeConfig(encodingType EncodingType) (merkletree.Config, error) {
+	switch encodingType {
+	case EncodingTypeBlindedSHA256v1:
+		return merkletree.NewConfig(SHA256Hasher{}, 2, MaxChildrenPerLeaf, EncodedLeaf{}), nil
+	case EncodingTypeBlindedSHA512_256v1:
+		return merkletree.NewConfig(SHA512_256Hasher{}, 2, MaxChildrenPerLeaf, EncodedLeaf{}), nil
+	}
+	return merkletree.Config{}, errors.Errorf("unknown encoding type %q", encodingType)
+}
+
+type EncodedLeaf []byte
+
+var _ merkletree.ValueConstructor = (*EncodedLeaf)(nil)
+
+func (l EncodedLeaf) Construct() interface{} {
+	return &[]byte{}
+}
+
+type LeafType uint16
+
+const (
+	LeafTypeChain17v1 = 1
+)
+
+type LeafContainer struct {
+	_struct   bool     `codec:",toarray"`
+	LeafType  LeafType // specifies structure of leafBytes
+	LeafBytes []byte   // msgpack deserialization implements Leaf
+}
+
+func NewLeafContainer(leafType LeafType, leafBytes []byte) LeafContainer {
+	return LeafContainer{LeafType: leafType, LeafBytes: leafBytes}
+}
+
+func (c LeafContainer) Serialize() ([]byte, error) {
+	return msgpack.EncodeCanonical(c)
+}
+
+type Leaf interface {
+	Serialize() ([]byte, error)
+	Type() LeafType
+	ID() []byte
+	GetSeqno() keybase1.Seqno
+}
+
+type Chain17v1Leaf struct {
+	_struct bool `codec:",toarray"`
+	TeamID  sig3.TeamID
```

Comment:
by the way, I think we can get rid of this...the msgpack includes the full key in the leaf's child map

Reply:
although it's only a constant space factor in the end

Reply:
oh that's a good point. I'd be in favor of shaving it off then